### PR TITLE
PXC-4288: Galera Arbitrator (garbd) uses 100% CPU

### PR DESCRIPTION
**Pull request by kamil-holubicki**
https://jira.percona.com/browse/PXC-4288

Problem:
Galera Arbitrator (garbd) uses 100% CPU

Cause:
PXC introduces P_S instrumentation for some of Galera mutexes and condition variables. For this, the library needs to be compiled with -DHAVE_PSI_INTERFACE=1.
As there is no PSI interface provider in the case of garbd, it has to be compiled without -DHAVE_PSI_INTERFACE=1.
Up to 8.0.32 we used scons for building Galera lib and garbd. Build process was done in 2 attempts:
1. Galera lib with -DHAVE_PSI_INTERFACE=1
2. garbd without -DHAVE_PSI_INTERFACE=1 PXC 8.0.33 switched  CMake to build Galera library and garbd and added -DHAVE_PSI_INTERFACE=1 globally, so both were compiled with this option. garbd registers and empty function in place of PSI instrumentation callback so effectively all mutexes and condition variables were noops. This caused observed 100% CPU utilization because of event processing loop turning around instead of waiting on a condition variable. Most probably other synchronization issues took place as well.

Solution:
Instead of defining separate CMake targets for garbd (gcomm, gcache, galerautilsxx) we always assume compilation with
-DHAVE_PSI_INTERFACE=1. Previously empty PSI callback was implemented to provide mutex and condition operations (similarly to wsrep_mysqld.cc wsrep_pfs_instr_cb function). So we use a kind of dependency injection pattern, where PSI instrumentation interface is always injected into Galera utils layer.
We still use plain Galera utils Mutex and Cond objects in case of compilation with GALERA_PSI_INTERFACE=OFF (Galera option) or WITH_PERFSCHEMA_STORAGE_ENGINE=OFF (server option) if

### CODE REVIEW

**Comment by kamil-holubicki** [comment]
https://pxc.cd.percona.com/job/pxc-8.0-pipeline-parallel-mtr/2008/
Failures are present also on 8.0:
```
galera.pxc_acf_add_delete_managed
galera.pxc_async_conn_failover
```

File: galerautils/src/gu_threads.h [hunk 1]
```diff
@@ -18,6 +18,7 @@ extern gu_pfs_instr_cb_t pfs_instr_callback;
 #if __unix__
 
 #include <pthread.h>
+#include <string.h>
```
Comment (dutow) [inline galerautils/src/gu_threads.h]: Is this change required?

Comment (kamil-holubicki) [reply galerautils/src/gu_threads.h]: I've reverted this change. Now it is the same as upstream.
This include is needed if `GU_DEBUG_MUTEX` is defined. Also, there is a problem in `Mutex::void disown()` method, as it references the unknown variable `value` instead of `value_`.
As I can not recall the case when we compiled Galera with GU_DEBUG_MUTEX I've reverted this change to not diverge from the upstream.

File: garb/garb_main.cpp [hunk 2]
```diff
@@ -224,14 +224,162 @@ main (int argc, char* argv[])
...
+        break;
+    }
+  }
+}
 
 int
 main (int argc, char* argv[])
 {
-    gu_conf_set_pfs_instr_callback(pfs_noop);
+    gu_conf_set_pfs_instr_callback(dummy_pfs_cb);
```
Comment (satya-bodapati) [inline garb/garb_main.cpp:382]: instead of providing dummy_pfs_cb, can we avoid registering this callback? what would happen? :thinking:

Comment (kamil-holubicki) [reply garb/garb_main.cpp:382]: Not a good idea. In case of compilation with `-DHAVE_PSI_INTERFACE=1` the surrounding code will call nullptr callback. And now by default we compile with `-DHAVE_PSI_INTERFACE=1` as it is needed by libgalera_smm.so (see changes in cmake files part)
In case of compilation without -DHAVE_PSI_INTERFACE=1, nothing SHOULD happen as the compiler should fall back to plain mutex/condition objects and this callback will be never called.
But prior to change we registered an empty callback always, so let's play safe.

Comment (satya-bodapati) [reply garb/garb_main.cpp:382]: agree

File: garb/garb_main.cpp [hunk 3]
```diff
@@ -224,14 +224,162 @@ main (int argc, char* argv[])
...
+   Galera lib with HAVE_PSI_INTERFACE defined and garbd with HAVE_PSI_INTERFACE
+   not defined. However, for garbd there is no "server side" which is PSI interface
+   provider (like for Galera lib).
+   The below function mocks PSI interface allowing garbd to work when compiled
+   with HAVE_PSI_INTERFACE defined. */
+static void dummy_pfs_cb(wsrep_pfs_instr_type_t type, wsrep_pfs_instr_ops_t ops,
+                        wsrep_pfs_instr_tag_t tag,
+                        void **value __attribute__((unused)),
+                        void **alliedvalue __attribute__((unused)),
+                        const void *ts __attribute__((unused))) {
```
Comment (venkatesh-prasad-v) [inline garb/garb_main.cpp]: Is `__attribute__((unused))` really required?

File: garb/garb_main.cpp [hunk 4]
```diff
@@ -224,14 +224,162 @@ main (int argc, char* argv[])
...
+        delete cond;
+        *value = nullptr;
+
+        break;
+      }
+
+      case WSREP_PFS_INSTR_OPS_WAIT: {
+        gu_cond_t *cond = reinterpret_cast<gu_cond_t *>(*value);
+        gu_mutex_t *mutex = reinterpret_cast<gu_mutex_t *>(*alliedvalue);
+        assert(cond != nullptr);
```
Comment (venkatesh-prasad-v) [inline garb/garb_main.cpp]:
```suggestion
        assert(cond != nullptr && mutex != nullptr);
```

File: garb/garb_main.cpp [hunk 5]
```diff
@@ -224,14 +224,162 @@ main (int argc, char* argv[])
...
+
+        gu_cond_destroy(cond);
+        delete cond;
+        *value = nullptr;
+
+        break;
+      }
+
+      case WSREP_PFS_INSTR_OPS_WAIT: {
+        gu_cond_t *cond = reinterpret_cast<gu_cond_t *>(*value);
```
Comment (venkatesh-prasad-v) [inline garb/garb_main.cpp:312]: May be having `assert(alliedvalue != nullptr);` is worth here?

File: garb/garb_main.cpp [hunk 6]
```diff
@@ -224,14 +224,162 @@ main (int argc, char* argv[])
...
+        gu_mutex_t *mutex = reinterpret_cast<gu_mutex_t *>(*alliedvalue);
+        assert(cond != nullptr);
+
+        gu_cond_wait (cond, mutex);
+
+        break;
+      }
+
+      case WSREP_PFS_INSTR_OPS_TIMEDWAIT: {
+        gu_cond_t *cond = reinterpret_cast<gu_cond_t *>(*value);
```
Comment (venkatesh-prasad-v) [inline garb/garb_main.cpp:322]: May be having `assert(alliedvalue != nullptr);` is worth here?

File: garb/garb_main.cpp (hunk 3 above)
Comment (venkatesh-prasad-v) [inline garb/garb_main.cpp]: Alignment
```suggestion
static void dummy_pfs_cb(wsrep_pfs_instr_type_t type, wsrep_pfs_instr_ops_t ops,
                         wsrep_pfs_instr_tag_t tag,
                         void **value __attribute__((unused)),
                         void **alliedvalue __attribute__((unused)),
                         const void *ts __attribute__((unused))) {
```

**Review by venkatesh-prasad-v — CHANGES_REQUESTED** [review]

File: garb/garb_main.cpp (hunk 4 above)
Comment (kamil-holubicki) [reply garb/garb_main.cpp]: done

File: garb/garb_main.cpp (hunk 5 above)
Comment (kamil-holubicki) [reply garb/garb_main.cpp:312]: adding assert for mutex is enough

File: garb/garb_main.cpp (hunk 6 above)
Comment (kamil-holubicki) [reply garb/garb_main.cpp:322]: having assert for mutex is enough

File: garb/garb_main.cpp (hunk 3 above)
Comment (kamil-holubicki) [reply garb/garb_main.cpp]: done

Comment (kamil-holubicki) [reply garb/garb_main.cpp]: This attribute, attached to a variable, means that the variable is meant to be possibly unused. GCC will not produce a warning for this variable.

Comment (venkatesh-prasad-v) [reply garb/garb_main.cpp]: Can we use the newer `[[maybe_unused]]` here?

**Review by venkatesh-prasad-v — APPROVED** [review]
I do not have any other comments other than the usage of [[maybe_unused]], which is optional. Looks good to me. Thank you for working on this issue.
